### DISCUSSION TITLE
Symlink to .colima/docker.sock from /var/run/docker.sock

### DISCUSSION THREAD

**Discussion by DeveloperJLLC** [abiosoft/colima#139] [Ideas]
A lot of applications connect to docker via /var/run/docker.sock see [testcontainers](https://github.com/testcontainers/testcontainers-java/blob/737865fb964f6fd5dd84ff6dbe8ca42ab9efdc53/docs/features/configuration.md
) jenkins might as well

It would be cool if the brew/apt install or startup could create a symlink to .colima/docker.sock from /var/run/docker.sock so that colima could be compatible with these applications. 

I already did this manually my self and it works using the following commands after installing colima

cd ~
colima start
cd /var/run
sudo ln -s ~/.colima/docker.sock docker.sock
cd ~

**Comment by abiosoft** [comment]
Initially, colima symlinks to the default docker socket path. This changed in v0.3.0 to ensure co-existence with other docker servers as well as supporting multiple instances.

You can always switch the docker context or set the `DOCKER_HOST` environment variable.

**Reply by DeveloperJLLC** [reply]
Yeah that works. Thanks for being curtious of other docker servers. Other docker servers would just make the symlink. A prompt to make colima the default would work as well.

**Comment by RahilRehan** [comment]
Hey, had the same issue yesterday with testcontainers-go. Exporting docker host worked for me.
`export DOCKER_HOST=unix:///Users/$USER/.colima/docker.sock instead of a symlink`

**Comment by rdeusser** [comment]
I understand the reasoning for wanting to work alongside other docker servers, but this essentially breaks the moby client as there isn't a method in the client to use the docker sock specified in the current context. That's all part of the docker CLI. So any projects that use the moby client instead of the docker CLI won't work out of the box unless you have an environment variable set which most people don't set because normally we're sure the docker sock is at `/var/run/docker.sock`. I'd just spent twenty minutes trying to figure out why `docker ps` works while the moby client didn't only to find out `/var/run/docker.sock` didn't actually exist.

**Comment by rdeusser** [comment]
And I'm not sure why yet, but even if I have the moby client explicitly connect to `~/.colima/docker.sock` after 10 minutes or so the socket stops working entirely. Even `docker ps` doesn't work. I don't know if there's a way for colima/lima to recreate that file without restarting the VM.

**Reply by abiosoft** [reply]
@rdeusser what version of colima are you running?

**Reply by rdeusser** [reply]
Here's the output from `colima version`:
```
colima version 0.5.2
git commit: 6b5b6fe0540e708f0c9d6e8919fab292c671fc72

runtime: docker
arch: aarch64
client: v20.10.21-rd
server: v20.10.20

kubernetes
Client Version: v1.24.6
Kustomize Version: v4.5.4
Server Version: v1.24.3+k3s1
```

**Reply by abiosoft** [reply]
Also, can you kindly share the output of `colima status`?

**Reply by rdeusser** [reply]
Sure can:
```
time="2023-01-23T16:14:10-05:00" level=info msg="colima is running using macOS Virtualization.Framework"
time="2023-01-23T16:14:10-05:00" level=info msg="arch: aarch64"
time="2023-01-23T16:14:10-05:00" level=info msg="runtime: docker"
time="2023-01-23T16:14:10-05:00" level=info msg="mountType: virtiofs"
time="2023-01-23T16:14:10-05:00" level=info msg="socket: unix:///Users/rdeusser/.colima/default/docker.sock"
time="2023-01-23T16:14:10-05:00" level=info msg="kubernetes: enabled"
```

**Reply by rdeusser** [reply]
I think I understand a little bit more about what's happening. So I had something that was creating a moby client, but for some reason the `defer client.Close()` wasn't working so it was rapidly creating clients and not closing them. This caused the docker sock to stop working _outside_ the VM, but _inside_ the VM it continued to work. Only thing I can think of is that the docker daemon crashed and restarted, but the link was severed between the host and the VM.
